### DISCUSSION TITLE
Custom TLV for IEEE 802.1 OUI (00-80-C2) Can Not Be Sent to Neighbor Side

### DISCUSSION THREAD

**Discussion by gozuacik** [lldpd/lldpd#532] [Q&A]
Hello,

We are running lldpd 1.0.11. We are trying to send a custom TLV for IEEE 802.1 OUI (00-80-C2) to Neighbor interface via llpdcli. However, neighbor interface is not updated with this TLV. What may cause this issue? Is there an additional parameter or configuration? Is this situation specific to IEEE 802.1 OUI (00-80-C2) ? For example, we can send another custom TLV with using a different OUI (01-80-C2)

_**Sender Side**_
```
[lldpcli] # configure ports eth1 lldp custom-tlv add oui 00,80,C2 subtype 18
[lldpcli] # show interfaces
-------------------------------------------------------------------------------
LLDP interfaces:
-------------------------------------------------------------------------------
Interface:    eth1
  Administrative status: RX and TX
  Chassis:     
    ChassisID:    mac dc:a6:32:c5:3c:c5
    SysName:      raspberrypi
    SysDescr:     Debian GNU/Linux 11 (bullseye) Linux 5.10.92-v8+ #1514 SMP PREEMPT Mon Jan 17 17:39:38 GMT 2022 aarch64
    MgmtIP:       192.168.1.100
    MgmtIface:    4
    MgmtIP:       fe80::518d:7941:fbd:1d69
    MgmtIface:    4
    Capability:   Bridge, off
    Capability:   Router, off
    Capability:   Wlan, on
    Capability:   Station, off
  Port:        
    PortID:       mac 00:e0:4c:04:02:58
    PortDescr:    eth1
  TTL:          120
  Unknown TLVs:
    TLV:          OUI: 00,80,C2, SubType: 18, Len: 0
-------------------------------------------------------------------------------
```

_**Receiver/Neighbor Side**_
```
lldpcli] # show neighbors
-------------------------------------------------------------------------------
LLDP neighbors:
-------------------------------------------------------------------------------
Interface:    enp0s10, via: LLDP, RID: 8, Time: 0 day, 00:04:49
  Chassis:     
    ChassisID:    mac dc:a6:32:c5:3c:c5
    SysName:      raspberrypi
    SysDescr:     Debian GNU/Linux 11 (bullseye) Linux 5.10.92-v8+ #1514 SMP PREEMPT Mon Jan 17 17:39:38 GMT 2022 aarch64
    MgmtIP:       192.168.1.100
    MgmtIface:    4
    MgmtIP:       fe80::518d:7941:fbd:1d69
    MgmtIface:    4
    Capability:   Bridge, off
    Capability:   Router, off
    Capability:   Wlan, on
    Capability:   Station, off
  Port:        
    PortID:       mac 00:e0:4c:04:02:58
    PortDescr:    eth1
    TTL:          120
-------------------------------------------------------------------------------
```

**Comment by vincentbernat** [comment]
Yes, the code assumes custom TLV are outside known OUI.

**Reply by gozuacik** [reply]
Thanks for quick response. But I couldn't get why lldpd/lldpcli is not able to send custom TLV for IEEE 802.1 OUI (00-80-C2)

**Reply by vincentbernat** [reply]
I think it is able to, but it will not receive them because it assumes custom TLV must not use known OUI (like 00-80-C2).

**Reply by gozuacik** [reply]
hmm, ok. In this case for known IEEE 802.1 OUI (00-80-C2), it seems we should not use custom TLV type. We would like to send Organizationally Specific TLV (#define LLDP_TLV_ORG 127) data for IEEE 802.1 OUI (00-80-C2).

**Reply by vincentbernat** [reply]
Check with tcpdump that lldpd can send the TLV. Then, we can make it accept it.

**Reply by gozuacik** [reply]
We checked that TLV data can be sent and received from tcpdump but it can not be read via lldpd. You can see the received tcpdump message below.

Sender side executes following command: _configure ports enp0s8 lldp custom-tlv add oui 00,80,C2 subtype 18 oui-info 01,00,08,15,2A,B6,15,21,C0,A8,00,0E,00_

Receiver side gets following message.

```
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 1500 bytes
08:40:04.648958 LLDP, length 232: CSRPxVDP
    0x0000:  0180 c200 000e 0800 27b4 6e1e 8100 0000  ........'.n.....
    0x0010:  88cc 0207 0408 0027 ccad c204 0703 0800  .......'........
    0x0020:  27b4 6e1e 0602 0078 0a08 4353 5250 7856  '.n....x..CSRPxV
    0x0030:  4450 0c65 5562 756e 7475 2031 382e 3034  DP.eUbuntu.18.04
    0x0040:  2e36 204c 5453 204c 696e 7578 2035 2e34  .6.LTS.Linux.5.4
    0x0050:  2e30 2d31 3234 2d67 656e 6572 6963 2023  .0-124-generic.#
    0x0060:  3134 307e 3138 2e30 342e 312d 5562 756e  140~18.04.1-Ubun
    0x0070:  7475 2053 4d50 2046 7269 2041 7567 2035  tu.SMP.Fri.Aug.5
    0x0080:  2031 313a 3433 3a33 3420 5554 4320 3230  .11:43:34.UTC.20
    0x0090:  3232 2078 3836 5f36 340e 0400 9c00 8010  22.x86_64.......
    0x00a0:  0c05 010a 0002 0f02 0000 0002 0010 1811  ................
    0x00b0:  02fe 8000 0000 0000 007f 235a 50bb 4bd1  ..........#ZP.K.
    0x00c0:  f602 0000 0002 0008 0665 6e70 3073 38fe  .........enp0s8.
    0x00d0:  0900 120f 0301 0000 0000 fe09 0012 0f01  ................
    0x00e0:  03ec 0100 1efe 11 00 80c2 1201 0008 152a  ...............*
    0x00f0:  b615 21c0 a800 0e00 0000                 ..!.......
1 packet captured
1 packet received by filter
0 packets dropped by kernel
```

By the way, when we checked IEEE Std 802.1CS™‐2020 (Link‐local Registration Protocol) standard document, in **C.2.1.3** IEEE 802.1 OUI section, OUI is defined as 00-80-C2. Again, according to specification, ECP and TCP discovery TLVs are presented as below.

![image](https://user-images.githubusercontent.com/19507070/184894996-62ade105-f34b-4bf7-a5bf-96bcf2971ca8.png)

![image](https://user-images.githubusercontent.com/19507070/184895240-3e8a0166-78af-4cea-abe0-87fe7af51047.png)

Based on our evaluation and understanding, current LLDP implementation does not give us the functionality to read ECP and TCP discovery TLVs with OUI 00-80-C2? Or if that is not true, how should we set and read ECP and TCP discovery TLVs with LLDP?

**Reply by vincentbernat** [reply]
09b95cb62f56 should fix that.

**Reply by gozuacik** [reply]
Thanks so much for the quick action & solution. We will give it a try. By the way, is there a planned date for a new release of lldpd?

**Reply by vincentbernat** [reply]
1.0.15 has just been released.

**Reply by gozuacik** [reply]
Hello Vincent, we tested 1.0.15 version with both lldpcli and internal project code. It seems working as we are now able to set TLVs with OUI 00-80-C2. Thanks for your help and support.
